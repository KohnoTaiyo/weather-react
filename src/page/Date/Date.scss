@use "../../styles/variables" as *;

.condition {
  display: flex;
  align-items: center;
  padding: 0 8px 8px 0;
  border-bottom: 1px solid $primaryColor;
  margin-bottom: 16px;

  &__name {
    font-size: $largeFontSize;
    font-weight: bold;
    margin-right: 16px;
  }
}

.status {
  &__text {
    display: flex;
    align-items: center;
    column-gap: 4px;

    &:not(:last-child) {
      margin-bottom: 8px;
    }
  }
}
