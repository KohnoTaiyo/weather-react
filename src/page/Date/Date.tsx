import "./Date.scss";

import { useEffect, useState } from "react";

import { BsFillSunriseFill, BsFillSunsetFill } from "react-icons/bs";
import { FaCloudRain, FaWind } from "react-icons/fa";
import { WiHumidity } from "react-icons/wi";
import { useLocation, useNavigate } from "react-router-dom";

import { TopSection } from "../../components/TopSection/TopSection";
import { getForecastWeather } from "../../hooks/fetchers/fetchers";
import { formatDate } from "../../hooks/formatDate";
import { ForecastFewDaysWeatherType, StateParams } from "../../types";

function Date() {
  const routerParams = useLocation();
  const pathState = routerParams.state as StateParams | undefined;
  const [text, setText] = useState(pathState?.location || "");
  const [errorMessage, setErrorMessage] = useState("");
  const [isLoading, setIsLoading] = useState(false);
  const [forecastWeather, setForecastWeather] = useState<ForecastFewDaysWeatherType | undefined>();
  const dayData = forecastWeather?.forecast.forecastday[0];
  const navigate = useNavigate();

  const fetchWeather = async () => {
    setIsLoading(true);
    try {
      const data = await getForecastWeather(text, 1, pathState?.date);
      setErrorMessage("");
      setForecastWeather(data);
    } catch (err) {
      if (err instanceof Error) {
        setErrorMessage(err.message);
      }
      setForecastWeather(undefined);
    } finally {
      setIsLoading(false);
    }
  };

  const onSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const sendPathParams: StateParams = { location: text };
    try {
      if (!text) throw new Error("Please enter a location.");
      setErrorMessage("");
      navigate("/", { state: sendPathParams });
    } catch (err) {
      if (err instanceof Error) {
        setErrorMessage(err.message);
      }
    }
  };

  useEffect(() => {
    if (!pathState?.location || !pathState.date) {
      navigate("/");
    }
    fetchWeather();
  }, []);

  return (
    <>
      <TopSection
        title={`${pathState?.location} at ${formatDate(pathState?.date || "")}`}
        value={text}
        onValueChange={setText}
        onSubmit={onSubmit}
      />
      {isLoading ? (
        <p>Loading...</p>
      ) : (
        !errorMessage &&
        dayData && (
          <>
            <div className="condition">
              <img src={dayData.day.condition.icon} alt="" />
              <p className="condition__name">{dayData.day.condition.text}</p>
              <div className="condition__temp">
                <p>{`High: ${dayData.day.maxtemp_c}°C`}</p>
                <p>{`Low: ${dayData.day.mintemp_c}°C`}</p>
              </div>
            </div>
            <div className="status">
              <p className="status__text">
                <WiHumidity />
                {`Humidity: ${dayData.day.avghumidity}%`}
              </p>
              <p className="status__text">
                <FaWind />
                {`Max wind kph: ${dayData.day.maxwind_kph}km/h`}
              </p>
              <p className="status__text">
                <FaCloudRain />
                {`Chance of precipitation: ${dayData.day.daily_chance_of_rain}%`}
              </p>
              <p className="status__text">
                <BsFillSunriseFill />
                {`Sunrise: ${dayData.astro.sunrise}`}
              </p>
              <p className="status__text">
                <BsFillSunsetFill />
                {`Sunset: ${dayData.astro.sunset}`}
              </p>
            </div>
            <div className="time">
              {dayData.hour.map((item) => (
                <div className="time__item" key={item.time}>
                  <p>{item.time.slice(-5)}</p>
                  <img src={item.condition.icon} alt={item.condition.text} />
                  <p>{`${item.temp_c}°C`}</p>
                  <p>{`${item.chance_of_rain}%`}</p>
                </div>
              ))}
            </div>
          </>
        )
      )}
    </>
  );
}

export default Date;
