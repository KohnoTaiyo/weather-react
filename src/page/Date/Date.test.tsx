import { render, screen } from "@testing-library/react";

import { mockForecastWeather } from "../../mock";
import Date from "./Date";

describe("Date", () => {
  test("renders correctly", () => {
    const { container } = render(<Date forecastWeather={mockForecastWeather} />);
    expect(screen.getByText(/Humidity:/i)).toBeInTheDocument();
    expect(screen.getByText(/Sunset:/i)).toBeInTheDocument();
    expect(screen.getByText(/00:00/i)).toBeInTheDocument();
    expect(screen.getByText(/23:00/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
