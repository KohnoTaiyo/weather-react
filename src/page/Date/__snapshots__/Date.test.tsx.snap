// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App init renders correctly 1`] = `
<div>
  <h1>
    New York at 12/01
  </h1>
  <form>
    <div
      class="textBox"
    >
      <input
        class="textBox__input"
        placeholder="location"
        required=""
        type="text"
        value="New York"
      />
      <input
        class="textBox__submit"
        tabindex="0"
        type="submit"
        value="Search"
      />
    </div>
  </form>
  <p>
    Loading...
  </p>
</div>
`;

exports[`App renders correctly fetch success 1`] = `
<div>
  <h1>
    New York at 12/01
  </h1>
  <form>
    <div
      class="textBox"
    >
      <input
        class="textBox__input"
        placeholder="location"
        required=""
        type="text"
        value="New York"
      />
      <input
        class="textBox__submit"
        tabindex="0"
        type="submit"
        value="Search"
      />
    </div>
  </form>
  <div
    class="condition"
  >
    <img
      alt=""
      src="//cdn.weatherapi.com/weather/64x64/day/143.png"
    />
    <p
      class="condition__name"
    >
      Mist
    </p>
    <div
      class="condition__temp"
    >
      <p>
        High: 3.2°C
      </p>
      <p>
        Low: -1.6°C
      </p>
    </div>
  </div>
  <div
    class="status"
  >
    <p
      class="status__text"
    >
      <svg
        fill="currentColor"
        height="1em"
        id="Layer_1"
        stroke="currentColor"
        stroke-width="0"
        version="1.1"
        viewBox="0 0 30 30"
        width="1em"
        x="0px"
        xmlns="http://www.w3.org/2000/svg"
        y="0px"
      >
        <path
          d="M7.56,17.19c0-0.88,0.24-1.89,0.72-3.03s1.1-2.25,1.86-3.31c1.56-2.06,2.92-3.62,4.06-4.67l0.75-0.72
	c0.25,0.26,0.53,0.5,0.83,0.72c0.41,0.42,1.04,1.11,1.88,2.09s1.57,1.85,2.17,2.65c0.71,1.01,1.32,2.1,1.81,3.25
	s0.74,2.16,0.74,3.03c0,1-0.19,1.95-0.58,2.86c-0.39,0.91-0.91,1.7-1.57,2.36c-0.66,0.66-1.45,1.19-2.37,1.58
	c-0.92,0.39-1.89,0.59-2.91,0.59c-1,0-1.95-0.19-2.86-0.57c-0.91-0.38-1.7-0.89-2.36-1.55c-0.66-0.65-1.19-1.44-1.58-2.35
	S7.56,18.23,7.56,17.19z M9.82,14.26c0,0.83,0.17,1.49,0.52,1.99c0.35,0.49,0.88,0.74,1.59,0.74c0.72,0,1.25-0.25,1.61-0.74
	c0.35-0.49,0.53-1.15,0.54-1.99c-0.01-0.84-0.19-1.5-0.54-2c-0.35-0.49-0.89-0.74-1.61-0.74c-0.71,0-1.24,0.25-1.59,0.74
	C9.99,12.76,9.82,13.42,9.82,14.26z M11.39,14.26c0-0.15,0-0.27,0-0.35s0.01-0.19,0.02-0.33c0.01-0.14,0.02-0.25,0.05-0.32
	s0.05-0.16,0.09-0.24c0.04-0.08,0.09-0.15,0.15-0.18c0.07-0.04,0.14-0.06,0.23-0.06c0.14,0,0.25,0.04,0.33,0.12s0.14,0.21,0.17,0.38
	c0.03,0.18,0.05,0.32,0.06,0.45s0.01,0.3,0.01,0.52c0,0.23,0,0.4-0.01,0.52c-0.01,0.12-0.03,0.27-0.06,0.45
	c-0.03,0.17-0.09,0.3-0.17,0.38s-0.19,0.12-0.33,0.12c-0.09,0-0.16-0.02-0.23-0.06c-0.07-0.04-0.12-0.1-0.15-0.18
	c-0.04-0.08-0.07-0.17-0.09-0.24c-0.02-0.08-0.04-0.19-0.05-0.32c-0.01-0.14-0.02-0.25-0.02-0.32S11.39,14.41,11.39,14.26z
	 M11.98,22.01h1.32l4.99-10.74h-1.35L11.98,22.01z M16.28,19.02c0.01,0.84,0.2,1.5,0.55,2c0.35,0.49,0.89,0.74,1.6,0.74
	c0.72,0,1.25-0.25,1.6-0.74c0.35-0.49,0.52-1.16,0.53-2c-0.01-0.84-0.18-1.5-0.53-1.99c-0.35-0.49-0.88-0.74-1.6-0.74
	c-0.71,0-1.25,0.25-1.6,0.74C16.47,17.52,16.29,18.18,16.28,19.02z M17.85,19.02c0-0.23,0-0.4,0.01-0.52
	c0.01-0.12,0.03-0.27,0.06-0.45s0.09-0.3,0.17-0.38s0.19-0.12,0.33-0.12c0.09,0,0.17,0.02,0.24,0.06c0.07,0.04,0.12,0.1,0.16,0.19
	c0.04,0.09,0.07,0.17,0.1,0.24s0.04,0.18,0.05,0.32l0.01,0.32l0,0.34c0,0.16,0,0.28,0,0.35l-0.01,0.32l-0.05,0.32l-0.1,0.24
	l-0.16,0.19l-0.24,0.06c-0.14,0-0.25-0.04-0.33-0.12s-0.14-0.21-0.17-0.38c-0.03-0.18-0.05-0.33-0.06-0.45S17.85,19.25,17.85,19.02z
	"
        />
      </svg>
      Humidity: 97%
    </p>
    <p
      class="status__text"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M156.7 256H16c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h142.2c15.9 0 30.8 10.9 33.4 26.6 3.3 20-12.1 37.4-31.6 37.4-14.1 0-26.1-9.2-30.4-21.9-2.1-6.3-8.6-10.1-15.2-10.1H81.6c-9.8 0-17.7 8.8-15.9 18.4 8.6 44.1 47.6 77.6 94.2 77.6 57.1 0 102.7-50.1 95.2-108.6C249 291 205.4 256 156.7 256zM16 224h336c59.7 0 106.8-54.8 93.8-116.7-7.6-36.2-36.9-65.5-73.1-73.1-55.4-11.6-105.1 24.9-114.9 75.5-1.9 9.6 6.1 18.3 15.8 18.3h32.8c6.7 0 13.1-3.8 15.2-10.1C325.9 105.2 337.9 96 352 96c19.4 0 34.9 17.4 31.6 37.4-2.6 15.7-17.4 26.6-33.4 26.6H16c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16zm384 32H243.7c19.3 16.6 33.2 38.8 39.8 64H400c26.5 0 48 21.5 48 48s-21.5 48-48 48c-17.9 0-33.3-9.9-41.6-24.4-2.9-5-8.7-7.6-14.5-7.6h-33.8c-10.9 0-19 10.8-15.3 21.1 17.8 50.6 70.5 84.8 129.4 72.3 41.2-8.7 75.1-41.6 84.7-82.7C526 321.5 470.5 256 400 256z"
        />
      </svg>
      Max wind kph: 10.4km/h
    </p>
    <p
      class="status__text"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M416 128c-.6 0-1.1.2-1.6.2 1.1-5.2 1.6-10.6 1.6-16.2 0-44.2-35.8-80-80-80-24.6 0-46.3 11.3-61 28.8C256.4 24.8 219.3 0 176 0 114.1 0 64 50.1 64 112c0 7.3.8 14.3 2.1 21.2C27.8 145.8 0 181.5 0 224c0 53 43 96 96 96h320c53 0 96-43 96-96s-43-96-96-96zM88 374.2c-12.8 44.4-40 56.4-40 87.7 0 27.7 21.5 50.1 48 50.1s48-22.4 48-50.1c0-31.4-27.2-43.1-40-87.7-2.2-8.1-13.5-8.5-16 0zm160 0c-12.8 44.4-40 56.4-40 87.7 0 27.7 21.5 50.1 48 50.1s48-22.4 48-50.1c0-31.4-27.2-43.1-40-87.7-2.2-8.1-13.5-8.5-16 0zm160 0c-12.8 44.4-40 56.4-40 87.7 0 27.7 21.5 50.1 48 50.1s48-22.4 48-50.1c0-31.4-27.2-43.1-40-87.7-2.2-8.1-13.5-8.5-16 0z"
        />
      </svg>
      Chance of precipitation: 0%
    </p>
    <p
      class="status__text"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.646 1.146a.5.5 0 0 1 .708 0l1.5 1.5a.5.5 0 0 1-.708.708L8.5 2.707V4.5a.5.5 0 0 1-1 0V2.707l-.646.647a.5.5 0 1 1-.708-.708l1.5-1.5zM2.343 4.343a.5.5 0 0 1 .707 0l1.414 1.414a.5.5 0 0 1-.707.707L2.343 5.05a.5.5 0 0 1 0-.707zm11.314 0a.5.5 0 0 1 0 .707l-1.414 1.414a.5.5 0 1 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zM11.709 11.5a4 4 0 1 0-7.418 0H.5a.5.5 0 0 0 0 1h15a.5.5 0 0 0 0-1h-3.79zM0 10a.5.5 0 0 1 .5-.5h2a.5.5 0 0 1 0 1h-2A.5.5 0 0 1 0 10zm13 0a.5.5 0 0 1 .5-.5h2a.5.5 0 0 1 0 1h-2a.5.5 0 0 1-.5-.5z"
        />
      </svg>
      Sunrise: 07:07 AM
    </p>
    <p
      class="status__text"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.646 4.854a.5.5 0 0 0 .708 0l1.5-1.5a.5.5 0 0 0-.708-.708l-.646.647V1.5a.5.5 0 0 0-1 0v1.793l-.646-.647a.5.5 0 1 0-.708.708l1.5 1.5zm-5.303-.51a.5.5 0 0 1 .707 0l1.414 1.413a.5.5 0 0 1-.707.707L2.343 5.05a.5.5 0 0 1 0-.707zm11.314 0a.5.5 0 0 1 0 .706l-1.414 1.414a.5.5 0 1 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zM11.709 11.5a4 4 0 1 0-7.418 0H.5a.5.5 0 0 0 0 1h15a.5.5 0 0 0 0-1h-3.79zM0 10a.5.5 0 0 1 .5-.5h2a.5.5 0 0 1 0 1h-2A.5.5 0 0 1 0 10zm13 0a.5.5 0 0 1 .5-.5h2a.5.5 0 0 1 0 1h-2a.5.5 0 0 1-.5-.5z"
        />
      </svg>
      Sunset: 04:28 PM
    </p>
  </div>
  <div
    class="time"
  >
    <div
      class="time__item"
    >
      <p>
        00:00
      </p>
      <img
        alt="Freezing fog"
        src="//cdn.weatherapi.com/weather/64x64/night/260.png"
      />
      <p>
        -0.3°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        01:00
      </p>
      <img
        alt="Freezing fog"
        src="//cdn.weatherapi.com/weather/64x64/night/260.png"
      />
      <p>
        -0.5°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        02:00
      </p>
      <img
        alt="Freezing fog"
        src="//cdn.weatherapi.com/weather/64x64/night/260.png"
      />
      <p>
        -0.5°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        03:00
      </p>
      <img
        alt="Freezing fog"
        src="//cdn.weatherapi.com/weather/64x64/night/260.png"
      />
      <p>
        -1°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        04:00
      </p>
      <img
        alt="Freezing fog"
        src="//cdn.weatherapi.com/weather/64x64/night/260.png"
      />
      <p>
        -1.4°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        05:00
      </p>
      <img
        alt="Mist"
        src="//cdn.weatherapi.com/weather/64x64/night/143.png"
      />
      <p>
        -1.5°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        06:00
      </p>
      <img
        alt="Mist"
        src="//cdn.weatherapi.com/weather/64x64/night/143.png"
      />
      <p>
        -1.6°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        07:00
      </p>
      <img
        alt="Mist"
        src="//cdn.weatherapi.com/weather/64x64/night/143.png"
      />
      <p>
        -1.4°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        08:00
      </p>
      <img
        alt="Mist"
        src="//cdn.weatherapi.com/weather/64x64/day/143.png"
      />
      <p>
        -1.6°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        09:00
      </p>
      <img
        alt="Sunny"
        src="//cdn.weatherapi.com/weather/64x64/day/113.png"
      />
      <p>
        5°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        10:00
      </p>
      <img
        alt="Mist"
        src="//cdn.weatherapi.com/weather/64x64/day/143.png"
      />
      <p>
        -0.8°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        11:00
      </p>
      <img
        alt="Mist"
        src="//cdn.weatherapi.com/weather/64x64/day/143.png"
      />
      <p>
        -0.1°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        12:00
      </p>
      <img
        alt="Mist"
        src="//cdn.weatherapi.com/weather/64x64/day/143.png"
      />
      <p>
        1.6°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        13:00
      </p>
      <img
        alt="Mist"
        src="//cdn.weatherapi.com/weather/64x64/day/143.png"
      />
      <p>
        2.4°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        14:00
      </p>
      <img
        alt="Mist"
        src="//cdn.weatherapi.com/weather/64x64/day/143.png"
      />
      <p>
        2.3°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        15:00
      </p>
      <img
        alt="Mist"
        src="//cdn.weatherapi.com/weather/64x64/day/143.png"
      />
      <p>
        3°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        16:00
      </p>
      <img
        alt="Mist"
        src="//cdn.weatherapi.com/weather/64x64/day/143.png"
      />
      <p>
        3°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        17:00
      </p>
      <img
        alt="Fog"
        src="//cdn.weatherapi.com/weather/64x64/night/248.png"
      />
      <p>
        3.2°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        18:00
      </p>
      <img
        alt="Fog"
        src="//cdn.weatherapi.com/weather/64x64/night/248.png"
      />
      <p>
        3.2°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        19:00
      </p>
      <img
        alt="Fog"
        src="//cdn.weatherapi.com/weather/64x64/night/248.png"
      />
      <p>
        3.5°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        20:00
      </p>
      <img
        alt="Fog"
        src="//cdn.weatherapi.com/weather/64x64/night/248.png"
      />
      <p>
        2.9°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        21:00
      </p>
      <img
        alt="Fog"
        src="//cdn.weatherapi.com/weather/64x64/night/248.png"
      />
      <p>
        3.7°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        22:00
      </p>
      <img
        alt="Fog"
        src="//cdn.weatherapi.com/weather/64x64/night/248.png"
      />
      <p>
        4.3°C
      </p>
      <p>
        0%
      </p>
    </div>
    <div
      class="time__item"
    >
      <p>
        23:00
      </p>
      <img
        alt="Fog"
        src="//cdn.weatherapi.com/weather/64x64/night/248.png"
      />
      <p>
        4.2°C
      </p>
      <p>
        0%
      </p>
    </div>
  </div>
</div>
`;
